### PR TITLE
Split Twig_Environment from TwigRendererFactory

### PR DESCRIPTION
**Pull request by geerteltink**
This PR splits the Twig_Environment from the TwigRendererFactory. It makes the Twig_Environment reusable where needed as discussed in zendframework/zend-expressive-twigrenderer#14.

The Twig_Environment must be registered with its TwigEnvironmentFactory.

```php
return [
    'dependencies' => [
        'factories' => [
            Twig_Environment::class                                   => Zend\Expressive\Twig\TwigEnvironmentFactory::class,
            Zend\Expressive\Template\TemplateRendererInterface::class => Zend\Expressive\Twig\TwigRendererFactory::class,
        ],
    ],

    // ...
];
```

- [x] Move Twig_Environment into it's own factory
- [x] Write tests
- [x] Test in a zend-expressive application
- [x] Update zend-expressive skeleton installer

### CODE REVIEW

**Comment by geerteltink** [comment]
@weierophinney I've tested this on a existing app and it seems to work. 

Just wondering if config can be added for these: ~~cache~~, debug (under the twig namespace), strict_variables and auto_reload.  Or should they go into the a separate pr?

**Comment by geerteltink** [comment]
@weierophinney I'm actually waiting for a code review. Since you were talking about using delegators, I want to make sure this is a good approach before I start rewriting the tests. I thought I already mentioned this but I didn't.

**Comment by glen-84** [comment]
@xtreamwayz @weierophinney -- Any news here? I really need a clean way of accessing the Twig_Environment.

**Comment by geerteltink** [comment]
Sorry, I'm waiting on a code review before I start rewriting all the tests.

In the meantime you could write your own factories or use the ones in this PR to get you started.

**Comment by glen-84** [comment]
@michaelmoussa Are you able to review this?

**Comment by michaelmoussa** [comment]
@glen-84 @xtreamwayz this looks fine to me - pretty straightforward extraction.

**Comment by geerteltink** [comment]
Ok, I'll update the tests probably this weekend.

**Comment by glen-84** [comment]
Will this make it into the next round of releases (with the query string stuff, etc.)?

File: src/TwigEnvironmentFactory.php
Comment (weierophinney) [inline src/TwigEnvironmentFactory.php:2]: Needs licence docblock; I can do that on merge.

File: src/TwigEnvironmentFactory.php
Comment (weierophinney) [inline src/TwigEnvironmentFactory.php:11]: Why is the same class imported under two different names?

File: src/TwigEnvironmentFactory.php
Comment (weierophinney) [inline src/TwigEnvironmentFactory.php:141]: Extract the above two conditionals to a new method, please.

File: src/TwigEnvironmentFactory.php
Comment (weierophinney) [inline src/TwigEnvironmentFactory.php:178]: Since the previous assignment is several lines above, no need to align the `=` operators.

**Review by weierophinney — APPROVED** [review]
I'll make the changes requested during merge; overall, it looks like a straight-forward extraction.

We'll need to update the Twig documentation as well; I'll create an issue and/or PR to do that after merging.

File: src/TwigEnvironmentFactory.php
Comment (geerteltink) [reply src/TwigEnvironmentFactory.php:11]: I probably copy pasted it from somewhere. Good catch tough. I guess the TwigLoader will do for both.

File: src/TwigRendererFactory.php
Comment (weierophinney) [inline src/TwigRendererFactory.php:31]: Hm... This may be a BC break, on further review.

If a user was previously using the `TwigRenderer` service and upgrades, the `TwigEnvironment` service is not registered, which will cause this factory to fail.

My thought is to do this:

```php
$environment = $this->getEnvironment($container);
```

Where `getEnvironment()` is defined as follows:

```php
public function getEnvironment(ContainerInterface $container)
{
    if ($container->has(TwigEnvironment::class)) {
        return $container->get(TwigEnvironment::class);
    }

    $environmentFactory = new TwigEnvironmentFactory();
    return $environmentFactory($container);
}
```

This will keep usage backwards compatible for existing users, while allowing them to add the factory to their configuration in order to be forwards compatible.

I can do that on merge.

File: src/TwigEnvironmentFactory.php
Comment (geerteltink) [reply src/TwigEnvironmentFactory.php:178]: Blame PhpStorm :D

File: src/TwigRendererFactory.php
Comment (geerteltink) [reply src/TwigRendererFactory.php:31]: Let me correct that: 
> This ~~may be~~ *is* a BC break...

The TwigEnvironment has to be added manually on an update. That's why I had version 2.0 in mind, together with Twig version 2 (which got released recently).

Comment (weierophinney) [reply src/TwigRendererFactory.php:31]: With the changes I propose above, however, it would not _need_ to be added immediately, but could be added later. We could even raise a deprecation notice if we get to the logic where we create the factory, to indicate that users should start registering the factory within their application configuration.

**Comment by weierophinney** [comment]
Thanks, @xtreamwayz\! I have incorporated the changes I suggested in my review, and merged to develop at this time.

**Pull request merged**
